### ISSUE TITLE
unpersist the kernel bundle (was: API to trigger kernel rebundling)

### ISSUE THREAD

**Issue by warner** [Agoric/Agoric#5679]
## What is the Problem Being Solved?

How should a chain upgrade cause a simultaneous kernel upgrade?

For POLA purposes, the swingset kernel is loaded into its own SES `Compartment`, and `importBundle()` is the simplest way to load arbitrary module graphs into SES compartments, which means the kernel source code must be collected into a Bundle first. This bundling can take several seconds, so to save time during application startup, we don't want to re-bundle the kernel on every application launch. Currently, we bundle the kernel exactly once, the first time the host application ever runs, when it calls `initializeSwingset()` (which starts from an empty database and initializes all of the empty tables and `0`-valued counters, etc). The bundle is stored into the database with `kvStore.set('kernelBundle', bundle)`, and read from the DB on every subsequent application launch (`importBundle(kvStore.get('kernelBundle'))`). (We manage the `lockdown` and `supervisor` bundles needed for the xsnap worker in the same way).

#4375 is a proposal to allow the host application to deliberately change which kernel bundle is used, through a new API on the `controller` object. Half of this proposal seems fine: storing the kernel bundle(s) in the general bundle table (next to vat bundles), and only storing the BundleID in the `kvStore`. However the external API feels at odds with the way chain operators expect to upgrade their validators.

Many of these validator operators use a version-management tool (["cosmovisor"](https://github.com/cosmos/cosmos-sdk/tree/main/cosmovisor) in conjunction with a standard "keep the process running" supervisor like `systemd`. The cosmovisor tool watches for governance votes that call for a wholesale replacement of the validator software (via upgrade plans written out by the `x/upgrade` module). Operators are responsible for paying attention to these votes and placing the new binaries in a special directory. The chain knows to halt at the approved block height if it is not running the correct (newer) software. The cosmovisor tool knows how to switch out symlinks at the right time, so that any new launches of the validator software will use the new version, rather than the old. Together, the old validator exits at the apointed time, the supervisor reacts to a stopped validator as usual (it attempts to relaunch the program), which cause the new validator version to be launched, which then passes the "am I running the right version after `blockHeightTrigger`" check, allowing it to keep running. The block height is chosen to allow enough time (perhaps a few days) for the operator to download and verify the new code before the cutover must happen.

If swingset re-bundled its kernel code on every application launch, this would work fine: the new validator version would include new kernel code (probably a whole copy of `agoric-sdk`, as installed into various `node_modules/` directories), all re-launches of the validator after that point would get the new kernel code instead of the old kernel code, and the resulting kernel behavior would change at the right moment. However 1: rebundling the kernel is uncomfortable slow to do on every app launch (and adds to the downtime caused by simple reboots), and 2: if we're at all worried about inconsistent kernel BundleIDs (does rebundling the same source code always create the same bundle/ID?), then rebundling on every reboot would make it worse (although I think we already depend upon consistent bundle IDs, because the `kvStore.set('kernelBundle')` is included in the initial crankHash/activityHash).

So swingset needs a signal that says "this is a good time to rebundle". It needs to happen at least every time the kernel code (or its dependencies) changes. And it really shouldn't happen until the validator is willing to accept a behavior change, i.e. when the chain is upgraded.

I'm thinking that the host application should supply this signal, in the form of a `controller.something(versionstring)` API call. Swingset will take the supplied string and compare it against a value stored in the DB. If they match, nothing happens. If they differ, swingset will rebundle the kernel, store the new bundle for use upon next reboot, and update the DB entry (so the next invocation won't cause a rebundling). The host application should call this at each application restart, before launching swingset (so the new bundle will be used for the subsequent kernel import). 

The host application source code should be modified to provide a different string every time the host application's packaging (`package.json`) is changed to request a different version of the swingset package.

The string could be the swingset version, or it could be a simple counter: anything that is easy to change whenever there's any chance that the swingset kernel code or its bundled dependencies might change.

### Alternatives

It is tricky for a package to introspect its own version number from it's packaging metadata (at least in the [languages](https://github.com/python-versioneer/python-versioneer) I'm most familiar with, perhaps `JSON.parse(fs.readFileSync(new URL('../../package.json', import.meta.url).pathname)).version` would work), and even that only gets you a change when the author of the package decides to make a release.

I expect we will be testing chain upgrades from one git revision of the monorepo to another, without necessarily doing a full release process (bumping version numbers) on all packages first. And I suspect other users of Swingset will have similar needs, where they want to update Swingset-using applications in a way (perhaps by applying new commits to a private fork of the swingset-bearing monorepo, or by using `patch-package` to modify the local `node_modules/` copy of `@agoric/swingset-vat`) that does not create a version bump in the swingset packaging metadata.

So relying upon swingset itself to know when the code has changed doesn't seem like a good idea.

On the other hand, it's a little weird to see a swingset version/etc in the downstream client (host application) code, at the call site which interacts with `controller`. And there's a danger of the author who updates the downstream application's packaging (to use a new swingset version) forgetting to update that call site to reflect the change.

### More general kernel upgrade tasks

We haven't really thought through the larger-scale lifetime of a swingset kernel. We expect kernel data structures to evolve, and the DB state created today may be very different than the state used six months from now. Sometimes a new swingset will be created (the DB initialized) with one version, and then later used by a different version, which needs to be capable of handling the older data.

This might take the form of discrete DB versions: we introduce a `kvStore.get('version')` key, and each time the kernel launches, it checks the value, and performs some sort of upgrade-in-place process if it differs (and then updates the key). This would not require any special care by the host application, but would introduce an unpredictable stall in launch time, depending upon when upgrades happen and how much data must be upgraded. It also requires diligence on the kernel authors, to update the expected value every time the DB schema (or semantics) changes. Thus far we've been pretty casual about kernel changes, because most instances are just running from the current repository trunk, and actual kernel upgrades are still in our future.

Instead, we might aim for fine-grained eternal upgrade: each kernel component that uses the DB should be prepared to see data from arbitrary points in the kernel's history. This requires a different sort of diligence: no discrete schema changes, but constant backwards-compatibility.

Or, we might use the "now is a good time to re-bundle the kernel" signal to double as a "now is a good time to see if your DB needs upgrading" signal. This might move the unpredictable-duration upgrade stall to at least happen at a predictable time, and it would make the host application's operator more involved in data migration activity.

## Description of the Design

With the scheme I'm currently favoring, the changes look like this:

* `initializeSwingset()` grows an option to provide an initial `kernelBundleSequence` string, which defaults to an empty string
* `initializeSwingset()` bundles the kernel sources, installs the bundle in the bundle table, stores the bundleID in `kvStore.set('kernelBundleID')`, and stores the provided/default `kernelBundleSequence` string into `kvStore.set('kernelBundleSequence')`
* we add `controller.maybeRebundleKernel(sequenceString)`, which compares `sequenceString` against the saved `kernelBundleSequence`, and if different: rebundle the kernel sources, install the new bundle, store the bundleID, decref the old bundleID, update `kernelBundleSequence`
* 
* our host application (cosmic-swingset) should add  `kernelBundleSequence: '0'` to the options it gives `initializeSwingset()`
* cosmic-swingset also adds a call to `controller.maybeRebundleKernel('0')` and a DB commit to the code in `launch-chain.js`, just before the call to `makeSwingsetController()`
* 
* we update the release process checklist/automation to increment both `'0'` strings just before we change `cosmic-swingset/package.json` to use a different version of `@agoric/swingset-vat`

## Security Considerations

No security considerations per se, but consensus considerations require very explicit transitions from one version of the source code to another. I *think* this would make the validator upgrade process pretty straightforward, but we should think through the release-process changes and decide if this seems robust enough to make sure the kernel is rebundled whenever necessary.

## Test Plan

Unit tests which invoke `maybeRebundleKernel()` can exercise the code, however since the kernel code won't change between bundles, I'm not sure how to confirm that it really worked. Perhaps we use automated tests to make sure it doesn't explode, but use a manual test to make sure the bundle ID actually changes. The manual test would deliberately modify a file in the source tree (by adding a newline to `kernel.js`, creating unstaged/uncommitted changes in `git status`) after making the first bundle and before making the second bundle. The developer running the test would need to revert those changes before continuing.

cc @michaelfig @mhofman @FUDCo for thoughts

**Comment by warner** [comment]
@michaelfig recommended that we just stop bundling the kernel at initialization time, and instead find a way to load the kernel directly from disk on each application restart. I think that means bundling the sources but not creating the archive, and loading the in-RAM bundle immediately.

This might cause application restart to take a little longer, but he cautions against premature optimization. I measured the `bundleSource()` time for the kernel at 1.8s on my 2021 MBP (M1-Pro CPU), but if the validator already takes several seconds to startup, this might not be a big deal, and perhaps there is a faster Endo API I should be using instead of `bundleSource()`/`importBundle()`. This will have a bigger effect on unit test times, where we can share the bundles among multiple tests in the same file.

The nice thing about this approach is that it would remove the need for a new controller API entirely. Host applications get kernel behavior that depends on the code on disk at the time of launch. Leave the code alone, and your behavior stays the same. Replace the code with something different (at application upgrade time), and now the behavior is different. It would also remove about 1.5MB from the kernelDB, since we no longer store the kernel bundle in the DB.

We'd need to address the "now is a good time to upgrade your DB schema" separately.

Incidentally, we decided that the other bundles (lockdown, xsnap supervisor, liveslots) *should* be bundled at initialization time, and stored in the DB, but that host applications which wish to upgrade their behavior should approach it like they do dynamic vat bundles and contracts. The host app would need to call `controller.installBundle()` with the new e.g. liveslots bundle. Then we add an API like `controller.setLiveslotsBundleID(newLiveslotsBundleID)` for the three special values (lockdown, supervisor, liveslots). These calls do something like `kvStore.set('liveslots', liveslotsBundleID)` (after checking that the bundleID actually exists), and increment a refcount on the bundle. Once changed, all subsequent new dynamic vats will get the new liveslots instead of the old one.

**Comment by warner** [comment]
With @kriskowal 's help, I was able to build a benchmark jig. On my 2021 MBP, `bundleSource(kernel.js)` takes 1.8s, and `importBundle(kernelBundle)` takes 156ms, for a total time of about 2s. We built a function tentatively named `importLocation2()` which takes an entry-point filename, does the same sort of module-finding and source-transformation steps as `bundleSource` (but avoids serialization or zipfile manipulation), and uses `importLocation` to import the code: this takes 1009ms.

The library we built to implement `importLocation2` was a mismash of private endo/compartment-mapper things, so it really wants to live in Endo proper, probably has a new package that subsumes `bundle-source`, `import-bundle`, and a few others. That's going to need some help from @kriskowal (or, more realistically, I'd just dump it all onto his plate :).

But a 1s difference is small enough that we can live without it for MN-1. So my plan is:
* at initialization time, build and store bundles for all the standard vats, and lockdown/supervisor, but *not* the kernel
* at application launch time (`createSwingsetController`), use `bundleSource(kernel.js)` followed by an immediate `importBundle()`, costing 2s
* a later performance improvement will replace that bundleSource/importBundle with a new `importLocation2` function, reducing the time to 1s

I'll continue to handle the `kernelBundles` option, so unit tests that call `createSwingsetController` a lot can still amortize that bundling time. But the actual kernel bundle might need to move from `initializeSwingset(.. initializationOptions.kernelBundles.kernel)` to `makeSwingsetController(.. runtimeOptions.kernelBundle)` so it will be available at the correct time.

This will make `agd` / `ag-solo` startup about 1.8s slower (we still pay the 156ms import time, but not the 1.8s bundling time), but gets us the new upgrade-friendly no-kernel-bundle-in-DB behavior we want. Later, the `importLocation2` will drop this by 1s to 1009ms. Unit tests should take the same time. The one-time `agd`/`ag-solo` initialization time (`agd init` instead of `agd start`, I think) will be 1.8s faster, since we're skipping the kernel bundling.
